lib/internal/process/execution.js

```diff
@@ -238,10 +246,120 @@ function readStdin(callback) {
   });
 }
 
+/**
+ *
+ * Adds the TS message to the error stack
+ * At the 3rd line of the stack, the message is added
+ * @param {Error} originalStack the stack to decorate
+ * @param {string} newMessage the message to add to the error stack
+ * @returns {void}
+ */
+function decorateCJSErrorWithTSMessage(originalStack, newMessage) {
+  const lines = StringPrototypeSplit(originalStack, '\n');
+  ArrayPrototypeSplice(lines, 3, 0, newMessage);
+  return ArrayPrototypeJoin(lines, '\n');
+}
+
+/**
+ *
+ * Wrapper of evalScript
+ *
+ * This function wraps the evaluation of the source code in a try-catch block.
+ * If the source code fails to be evaluated, it will retry evaluating the source code
+ * with the TypeScript parser.
+ *
+ * If the source code fails to be evaluated with the TypeScript parser,
+ * it will rethrow the original error, adding the TypeScript error message to the stack.
+ *
+ * This way we don't change the behavior of the code, but we provide a better error message
+ * in case of a typescript error.
+ */
+function evalTypeScript(name, source, breakFirstLine, print, shouldLoadESM = false) {
+  try {
+    evalScript(name, source, breakFirstLine, print, shouldLoadESM);
+  } catch (originalError) {
+    try {
+      const strippedSource = stripTypeScriptModuleTypes(source, name, false);
+      evalScript(name, strippedSource, breakFirstLine, print, shouldLoadESM);
+      // Emit the experimental warning after the code was successfully evaluated.
+      emitExperimentalWarning('Type Stripping');
+    } catch (tsError) {
+      if (tsError.code === 'ERR_INVALID_TYPESCRIPT_SYNTAX') {
+        originalError.stack = decorateCJSErrorWithTSMessage(originalError.stack, tsError.message);
```

Comment:
We could also have a test for `eval("throw Object.defineProperty(new Error, \"stack\", { set() {throw this} })")`
```suggestion
        try {
          originalError.stack = decorateCJSErrorWithTSMessage(originalError.stack, tsError.message);
        } catch { /* Ignore potential errors coming from `stack` getter/setter */ }
```